File: app/src/main/java/com/norvera/confession/ui/main/MainViewModelFactory.java
package com.norvera.confession.ui.main;

import android.annotation.SuppressLint;
import android.content.Context;

import com.norvera.confession.data.AppRepository;
import com.norvera.confession.utils.InjectorUtils;

import androidx.annotation.NonNull;
import androidx.annotation.VisibleForTesting;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

public class MainViewModelFactory extends ViewModelProvider.NewInstanceFactory{

    @SuppressLint("StaticFieldLeak")
    private static volatile MainViewModelFactory INSTANCE;

    private final AppRepository mAppRepository;

    public static MainViewModelFactory getInstance(Context context) {

        if (INSTANCE == null) {
            synchronized (MainViewModelFactory.class) {
                if (INSTANCE == null) {
                    INSTANCE = new MainViewModelFactory(
                            InjectorUtils.provideAppRepository( InjectorUtils.provideDatabase(context), InjectorUtils.provideAppExecutor()));
                }
            }
        }
        return INSTANCE;
    }


    @VisibleForTesting
    public static void destroyInstance() {
        INSTANCE = null;
    }


    private MainViewModelFactory(AppRepository repository) {
        mAppRepository = repository;
    }

    // Note: This can be reused with minor modifications
    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
        //noinspection unchecked
        return (T) new MainViewModel( mAppRepository);
    }
}

File: app/src/main/java/com/norvera/confession/viewmodels/MainViewModelFactory.kt
package com.norvera.confession.viewmodels

import android.annotation.SuppressLint
import android.content.Context
import androidx.annotation.VisibleForTesting
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.norvera.confession.data.AppRepository
import com.norvera.confession.utils.InjectorUtils

class MainViewModelFactory private constructor(private val mAppRepository: AppRepository) :
    ViewModelProvider.NewInstanceFactory() {

    // Note: This can be reused with minor modifications
    override fun <T : ViewModel> create(modelClass: Class<T>): T {

        return MainViewModel(mAppRepository) as T
    }

    companion object {

        @SuppressLint("StaticFieldLeak")
        @Volatile
        private var INSTANCE: MainViewModelFactory? = null

        fun getInstance(context: Context): MainViewModelFactory? {

            if (INSTANCE == null) {
                synchronized(MainViewModelFactory::class.java) {
                    if (INSTANCE == null) {
                        INSTANCE =
                            MainViewModelFactory(
                                InjectorUtils.provideAppRepository(
                                    InjectorUtils.provideDatabase(context),
                                    InjectorUtils.provideAppExecutor()
                                )
                            )
                    }
                }
            }
            return INSTANCE
        }


        @VisibleForTesting
        fun destroyInstance() {
            INSTANCE = null
        }
    }
}
